waiter/src/waiter/token.clj

```diff
@@ -719,3 +736,117 @@
+                                  [(make-index-event "UPDATE" token-index-entry)
+                                   (assoc-in current-state [:token->token-index token] token-index-entry)]
+                                  [(make-index-event "DELETE" {:owner owner :token token})
+                                   (assoc current-state :token->token-index (dissoc token->token-index token))]))]
+                          (if event
+                            (do
+                              (async/>! tokens-event-chan event)
+                              [next-state watch-refresh-timeout-chan])
+                            [current-state watch-refresh-timeout-chan])))
+
+                      tokens-watch-channels-update-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "process-tokens-channels-update")
+                        (let [{:keys [channel-event watch-chan]} msg
+                              next-state (case channel-event
+                                           :add (do
+                                                  (async/put! watch-chan (or (vals token->token-index) '()))
+                                                  (async/tap tokens-event-mult watch-chan)
+                                                  (update current-state :watches-count inc))
+                                           :remove (update current-state :watches-count dec)
+                                           (throw (ex-info "Invalid tokens-watch-channels-update-chan event" {:event msg})))]
+                          (log/info "token-watch-maintainer watch count changed!" {:previous-count watches-count
+                                                                                    :new-count (get next-state :watches-count)})
+                          [next-state watch-refresh-timeout-chan]))
+
+                      watch-refresh-timeout-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "token-watch-maintainer" "refresh")
+                        (let [next-token->index (get-token->token-index-fn)
+                              [only-old-indexes only-next-indexes _] (data/diff token->token-index next-token->index)]
+                          ; if token in old-indexes and not in only-next-indexes, then those token indexes were deleted
+                          (doseq [[token {:keys [owner]}] only-old-indexes]
+                            (when-not (contains? only-next-indexes token)
+                              (async/>! tokens-event-chan (make-index-event "DELETE" {:owner owner :token token}))))
+                          ; if token in only-next-indexes, it has been updated (soft-delete included)
+                          (doseq [[token _] only-next-indexes]
+                            (async/>! tokens-event-chan (make-index-event "UPDATE" (get next-token->index token))))
```

Comment:
Please consider sending a single message aggregating the updates and deletes on `tokens-event-chan`.
The handler can break this message down into individual pieces if it wants to.